Add keyboard shortcuts explainer

dmsnell opened the pull request:
Follows #603
Resolves #700
Resolves #798
Resolves #985
Resolves #1167
Resolves #1222

This application has had keyboard shortcuts since #603 but they have
been undocumented making them hard to discover and remember.

In this patch we're picking up the work started in #700 and adding
a keyboard shortcut overylay, opened through the Help menu or by
pressing `CmdOrCtrl` + `?`. This dialog shows a list of the
app-specific keyboard shortcuts and what they do.

Oh yeah, and we're changing the "Find" shortcut to <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>F</kbd>
in accordance with earlier design decisions to differentiate "search within
this note" vs. "search all notes" and also removes the <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>Up</kbd>/<kbd>Down</kbd>
from note navigation so that it stops interfering with text selection.

<img width="648" alt="Screen Shot 2020-04-01 at 10 30 54 PM" src="https://user-images.githubusercontent.com/5431237/78214086-2f2ba900-7469-11ea-9035-6aa11a3c90f0.png">

![keyboardExplainer mov](https://user-images.githubusercontent.com/5431237/78214582-74040f80-746a-11ea-9ab1-48ddc062a11a.gif)

![Screen Shot 2020-04-16 at 7 42 38 PM](https://user-images.githubusercontent.com/5431237/79526592-7776c980-801a-11ea-918c-637a0db6aa4f.png)

## Testing

Run through the testing checklist.
Pay particular attention to keyboard shortcuts.
Try to break them and use them in unintended circumstances.

SylvesterWilmott commented:
This is great!

codebykat commented:
Very cool! I have a few comments:

* The height is a bit awkward (it's just... sliiiiiightly too tall for the content). Especially on browser it looks ugly that it winds up with a scrollbar.

<img width="1041" alt="Screen Shot 2020-04-02 at 5 14 37 PM" src="https://user-images.githubusercontent.com/52152/78311365-7a939500-7505-11ea-9843-b06ab27ce1c7.png">

* Loooooots of these don't actually work in the browser, so we should fix them or consider removing them.

* It's also not ideal that there's no way to view this dialog in the browser without already knowing the magical key combination. In related news, the key combo itself is not in the dialog? If you're in the app and open it via the menu, useful info -- and especially useful info because now you know how to open it in the browser, in fact the only way to open it in the browser.

* A few of them work with either Cmd or Ctrl (Mac/Firefox), which is odd.

codebykat commented:
Things that don't work in Firefox/OSX:

`ctrl + shift + F` appears to highlight the next character. `cmd +shift + f` is full screen
`ctrl +`, `ctrl -`, and `ctrl 0` do nothing -- with `cmd` instead of `ctrl`, they do adjust font size in the browser
`ctrl + N` goes to next line in editor, `cmd + N` is a new browser window
`ctrl + P` does nothing, `cmd + P` is the browser print dialog
`ctrl + shift + C` and `cmd + shift + C` do nothing
`ctrl + shift + N` obviously only does anything if you're in a small window

dmsnell commented:
@codebykat I have reviewed all the shortcuts and made some adjustments. I'll post more later, but I tested in Firefox and Safari and Electron.

dmsnell commented:
> It's also not ideal that there's no way to view this dialog in the browser without already knowing the magical key combination. In related news, the key combo itself is not in the dialog?

Yeah I don't know what to do about discoverability but I did add it to the list. Maybe we can put it somewhere in the note list or in settings?

> A few of them work with either Cmd or Ctrl (Mac/Firefox), which is odd.

I believe that's the legacy of the cross-platform support. Any ideas? We can probably check `isMacOS` where we listen for these but it seemed fine to allow some hotkeys to work across all platforms.

> Loooooots of these don't actually work in the browser, so we should fix them or consider removing them.

Should be fixed as of 7b90418 - I removed tips from the browser that only work in Electron and removed keys already handled by the browser, like font zooming and printing.

dmsnell commented:
We need some more hotkeys…

## Essential

 - Open trash view
 - Open all notes view
 - Navigate tags in tag list

## Nice to have

 - Restore note
 - Delete forever
 - Pin note
 - Toggle Markdown Enabled

belcherj commented:
Are all of Kat's comments addressed? Is it time for another review here?

dmsnell commented:
@belcherj they are all addressed and yes it's time for a new review.

I expect some discussion about how I addressed them but I tended to each point in the way that seemed best as I was doing it

dmsnell commented:
> Delete note shortcut doesn't seem to work anywhere, with any modifier key. Tested in Electron and Firefox.

Updated in a816e4d to accept backspace or delete.

> The shortcut to open this is given as modifier + ?, but it's actually modifier + / 

Used <kbd>/</kbd> in 6cd7beb

> Settings isn't dismissed by hitting the same hotkey again, only by ESC. However the keyboard shortcuts panel is toggled by its hotkey, so I think Settings should exhibit the same behavior.

I'd like to hold off on this for now due to questions about the change. It's not hard, but I'm not sure it's a positive change.

> The keyboard shortcuts panel stays open if you open e.g. Settings on top of it; it also allows you to toggle navigation elements in the background (e.g. "toggle tag list"). That seems basically fine but wanted to make sure it was intentional.

It was intentional to allow opening the keyboard shortcuts from anyway. Basically it was the refactoring of the dialog logic that gave me the idea. It wasn't intentional to make it non-modal, but I don't know if that's something we want to tackle here either. I think it would take a lot of work to prevent other interactions and I'm not sure they are problematic in the first place. Thoughts?

> In Electron it's showing my modifier key as CMD, in Firefox as CTRL. In general this is fine because that's the one that, indeed, works with the most combinations. There are still some inconsistencies between the two but those are mostly fine.

Any idea how to resolve this? I left some permissive, where either modifier works, so that people jumping between platforms can use their learned keyboard shortcuts. Balancing that out I didn't want to avoid the platform norms for the sake of consistency in our app.

> The weirdest one is "insert checklist" which ONLY works with CMD in Electron, but works with both in Firefox. It seems like it's the only one that's not a browser-specific behavior (i.e. print, font size, other stuff that is in menus) that isn't working with both modifiers.

Good catch that's apparently a result of trapping the key through Electron vs. in the app itself. Replaced Electron keybinding in 891b65c

> Should we also include TAB / SHIFT+TAB to indent / unindent list items?

They have their own issues and I'm nervous to broadcast that functionality even more right now because it takes about a paragraph to explain that you have to have a list bullet with an empty line after it in order for those to work.

dmsnell commented:
> As far as I'm concerned this is good to go whenever you feel like it's ready.

Thanks for all the review. It made the outcome here much better than when it started. With all the remaining deficiencies I'm going to merge. I'd like to continue iterating on keyboard functionality to make some quirks better, but this PR accomplishes quite a bit already:
 - improve consistency of keybindings across the platforms and in consistency with the native platform on which they run
 - add tooltips for shortcuts on buttons that have corresponding shortcuts
 - add keybinding map with menu option / sidebar indicator
 - fix broken shortcuts
